How tos - DOC-314 (#472)
matter-labs/zksync-web-era-docs: docs/dev/how-to/estimate-gas.md

# Estimate gas

## L1 transactions

> To estimate gas for L1 transactions, use the `eth_estimateGas` method.

Author: This first section is talking about L1<->L1, for completeness ... I'm making it clearer

    :::tip Important
    This value is often referred to as **limit, or gas limit, or L2 gas limit** in our documented examples. 
    :::

    3.1 Apply an alias to the addresses in the request if the sender address is a contract. If the sender is an EOA, no aliasing is required.

4. Get the base cost by calling the [`l2TransactionBaseCost`]((https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/interfaces/IMailbox.sol#L129) ) function with:
    - The gas price returned at step 2.
    - The minimum gas value returned at step 3. 
    - The L1 to L2 gas per pubdata byte constant that represents the amount of L2 gas required to send a single byte of data. 

```solidity
function l2TransactionBaseCost(
        uint256 _gasPrice,
        uint256 _l2GasLimit,
        uint256 _l2GasPerPubdataByteLimit
    ) external view returns (uint256);
```

The resulting base cost is given in wei for the L2 part of the transaction.

> 5. Send the transaction, including both the gas price and base cost in the value parameters, by calling the [`requestL2Transaction`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/interfaces/IMailbox.sol#L119) function. Include the gas limit value from step 3 in the `calldata`.

Author: OK amended, I also added a bit more info about the params.

```solidity
(function requestL2Transaction(
    address _contractL2,
    uint256 _l2Value,
    bytes calldata _calldata,
    uint256 _l2GasLimit,
    uint256 _l2GasPerPubdataByteLimit,
    bytes[] calldata _factoryDeps,
    address _refundRecipient

> ) external payable returns (bytes32 txHash); IzkSync interface??)

Author: I have copied exactly what's in the link instead, so the IzkSync reference is now removed. I'll let you know when I'm done going through and you can have a look.

const gasPrice = await wallet.providerL1!.getGasPrice();

// The calldata can be encoded in the same way as for Ethereum
const calldata = "0x...";
const gasLimit = BigNumber.from(1000);
const gasPerPubdataByte = BigNumber.from(800);

const txCostPrice = await wallet.getBaseCost({
  gasPrice,
  gasLimit,
  gasPerPubdataByte
});

console.log(`Executing the transaction will cost ${ethers.utils.formatEther(txCostPrice)} ETH`);

// Initiate L2 transfer via L1 and execute from zkSync Era wallet
const someL2Contract = "0x19a5bfcbe15f98aa073b9f81b58466521479df8d";
const executeTx = await wallet.requestExecute({
  calldata,
  l2GasLimit: gasLimit,
  gasPerPubdataByte,
  contractAddress: someL2Contract,
  overrides: {
    gasPrice,
    value: txCostPrice,
  },
});

await executeTx.wait();
```
@tab Python
```sh
// In progress. Check back later.
```
@tab Go
```sh
// In progress. Check back later.
```
@tab Java
```sh
// In progress. Check back later.
```
:::

### Result

6. The return value is a 256-bit unsigned integer in hexadecimal representing how much gas the transaction uses.

```json
{
    "jsonrpc": "2.0",
    "result": "0x25f64db",
    "id": 2
}
```

> 

Author: I'm leaving an extra space in between main sections for author readability.

## L2 transactions

Gas estimation for L2 to L2 transactions on zkSyncEra works in the same way as in Ethereum.

Supply the same struct as for a L1 to L2 transaction.

> ?? // can you add some code here please

Author: OK, the question is, for L2 to L2 transactions, should we use this method instead: https://era.zksync.io/docs/api/api.html#zks-estimatefee ? And, if so, why?
Also,  does `eth_estimateGas` still work as before for L2 to L2?

> ?? // can you add some code here please

Reviewer: `eth_estimateGas` works fine. 

`zks_estimateFee` will need to be used, but not now

> ?? // can you add some code here please

Author: Thank you :)

## L2 to L1 transactions

Two transactions are required: 

- An L2 transaction which sends a message of arbitrary length. 
- When the block the message resides in is proved on L1, verification and confirmation takes place.

For example, an L2 transaction uses a bridge to transfer funds. The tokens are removed from L2, and a message is sent to L1 saying the tokens were burned. On L1, the block containing the message is proved, and the bridge counterpart verifies that the token was indeed present in a certain block on L2 and supplies the funds.

1. Import the zkSync Era library or contract containing the required functionality. 

2. Get a `Contract` object that represents the [`Messenger`](../developer-guides/system-contracts.md#l1messenger) contract.

3. Transform the request into a raw bytes array.

4. Use the [`sendToL1`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l2/system-contracts/interfaces/IL1Messenger.sol#L5) function from the `IL1Messenger.sol` interface, passing the message as a raw bytes array.

```solidity
function sendToL1(bytes memory _message) external returns (bytes32);
```

:::warning What is a message?

> - A message functions like an event on Ethereum. The difference is that a message publishes data on L1. 

Reviewer: ```suggestion
- A message functions similar to event on Ethereum. The difference is that a message publishes data on L1. 
```

I think they are not like each other but rather have some similarities

Author: "like" is equivalent here, if you use "similar to" you have to say .... "A message functions in a similar way to an event" and that's a lot of extra words :)

- Its Solidity representation:
    ```solidity
    struct L2Message {
            address sender;
            bytes data;
            uint256 txNumberInblock;
    }
    ```
:::

5. The [`proveL2MessageInclusion`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/facets/Mailbox.sol#L46) function returns a boolean parameter indicating whether the message was sent successfully to L1.

```solidity
function proveL2MessageInclusion(
    uint256 _blockNumber,
    uint256 _index,
    L2Message memory _message,
    bytes32[] calldata _proof

> 

Reviewer: ```suggestion
```

) public view returns (bool) {
    return _proveL2LogInclusion(_blockNumber, _index, _L2MessageToLog(_message), _proof);   
}
```

### Example

::: code-tabs
@tab Solidity
```solidity
//SPDX-License-Identifier: Unlicense
pragma solidity ^0.8.0;

// Importing zkSync contract interface
import "@matterlabs/zksync-contracts/l1/contracts/zksync/interfaces/IZkSync.sol";

contract Example {

>     function sendMessageToL1() external returns(bytes32 messageHash) {

Reviewer: This function should be present only on L2, it may be a bit confusing that the `Example` contract contains both L1 and L2 functions

Author: We say in the beginning of this section:

"Two transactions are required: 

- An L2 transaction which sends a message of arbitrary length. 
- When the block the message resides in is proved on L1, verification and confirmation takes place."

So there is one on L2, then one on L1 ... ?? We may need a quick chat :)

Author: OK, so the question is, when we say **two transactions** are required, are we also saying that these transactions cannot be in the same contract?

Author: we can reconvene on this one ..

        // Construct the message directly on the contract
        bytes memory message = abi.encode(address(this));

        messageHash = L1_MESSENGER_CONTRACT.sendToL1(message);
    }

  // NOTE: The zkSync contract implements only the functionality for proving that a message belongs to a block
  // but does not guarantee that such a proof was used only once. That's why a contract that uses L2 -> L1
  // communication must take care of the double handling of the message.
  /// @dev mapping L2 block number => message number => flag
  /// @dev Used to indicated that zkSync L2 -> L1 message was already processed
  mapping(uint256 => mapping(uint256 => bool)) isL2ToL1MessageProcessed;

  function consumeMessageFromL2(
    // The address of the zkSync smart contract.

>     // It is not recommended to hardcode it during the alpha testnet as regenesis may happen.

Reviewer: regenesis is unlikely, but yeah, it is better not to hardcode it

Author: we are out of alpha testnet, so I will remove this line

 * 3. Wait for block to be verified and validate proof
 */
async function main() {
  // Step 1: send message
  const l2Trx = await sendMessageToL1(MESSAGE);

  console.log("Waiting for transaction to finalize...");

  // Step 2: waiting to finalize can take a few minutes.
  const l2Receipt = await l2Trx.waitFinalize();

  // Step 3: get and validate proof (block must be verified)
  const proof = await getL2MessageProof(l2Receipt.blockNumber);

  console.log(`Proof is: `, proof);

  const { l1BatchNumber, l1BatchTxIndex } = await l2Provider.getTransactionReceipt(l2Receipt.transactionHash);

  console.log("L1 Index for Tx in block :>> ", l1BatchTxIndex);

  console.log("L1 Batch for block :>> ", l1BatchNumber);

  // IMPORTANT: This method requires that the block is verified
  // and sent to L1!
  const result = await proveL2MessageInclusion(
    l1BatchNumber,
    proof,
    // @ts-ignore
    l1BatchTxIndex
  );

  console.log("Result is :>> ", result);
  process.exit();
}

try {
  main();
} catch (error) {
  console.error(error);
}
```
@tab Python
```sh
// In progress. Check back later.
```
@tab Go
```sh
// In progress. Check back later.
```
:::

### Result

> 5. The return value is the `keccak256` hash of the message bytes.

Reviewer: `proveL2MessageInclusion` returns `true` or `false` depending on whether the proof is correct. (if that's what you mean by `result` here

Author: Thanks :) I think I probably need to rework the examples a little bit.

Author: I have amended the returns in this section.

# Estimate gas

> 

Author: then we are working backwards alphabetically ... this page has changed dramatically now so I'm going to close this comment... let's chat later... this is going to take some serious thought and is worth taking time over ... i'll let you know when i'm ready for a review ...

## L1 to L1 transactions

> 

Author: for completeness ... L1 to L1 for consistency with rest of page and to avoid people assuming layer interops ...

> To estimate gas for L1 to L1 transactions, use the `eth_estimateGas` method.

Reviewer 2: the important bit -- this eth_estimateGas should be sent do some L1 server - not to zksync, right?

Author: correct

Author: will deal with this

# Estimate gas

:::tip Gas fees
- Layer 2 gas fees depend on the current Ethereum L1 gas fees for publishing and verification.
:::

Gas estimation is usually left to the SDK to deal with. However, there are situations in which a developer may wish to implement gas estimation directly; such as for displaying the fee to the user for example.

This document gives step-by-step instructions for implementing gas estimation for:

- [L1 to L1](#l1-to-l1) transactions.
- [L1 to L2](#l1-to-l2) transactions.
- [L2 to L2](#l2-to-l2) transactions.

## L1 to L1

To estimate gas for L1 to L1 transactions, use the Ethereum `eth_estimateGas` method and send to an L1 node.

:::info More info
For more information and live testing, check out the [Ethereum JSON RPC docs](https://ethereum.github.io/execution-apis/api-documentation/).
:::

## L1 to L2

To estimate gas for an L1 to L2 transaction, first gather the required values:

1. Get the current L1 gas price. 
    
    For example, the Ethers.js library has a [`getGasPrice()`](https://docs.ethers.org/v5/api/providers/provider/#Provider-getGasPrice) method available on a `Provider` object. Your language of choice should have an equivalent.

> 2. Call the [`zks_estimateGasL1ToL2`](../../api/api.md#zks_estimategasl1tol2) method, passing the transaction data. This returns the minimum amount of gas the transaction requires; commonly called **gas limit** or similar in our code and docs.

Author: That is not what I was told. Please advise @StanislavBreadless .. also, I'm going to leave in the commonly called bit because it was not clear that these variable names were referring to a gas estimation limit (minimum or maximum) ...

> 2. Call the [`zks_estimateGasL1ToL2`](../../api/api.md#zks_estimategasl1tol2) method, passing the transaction data. This returns the minimum amount of gas the transaction requires; commonly called **gas limit** or similar in our code and docs.

Author: I'll wait for Stas response and then commit the change ..

> 2. Call the [`zks_estimateGasL1ToL2`](../../api/api.md#zks_estimategasl1tol2) method, passing the transaction data. This returns the minimum amount of gas the transaction requires; commonly called **gas limit** or similar in our code and docs.

Reviewer 3: > also, I'm going to leave in the commonly called bit because it was not clear that these variable names were referring to a gas estimation limit (minimum or maximum) ...

estimateGas should return the estimated amount gas to be consumed by the transaction. It should not be "gas limit" (whether min or max)

> 2. Call the [`zks_estimateGasL1ToL2`](../../api/api.md#zks_estimategasl1tol2) method, passing the transaction data. This returns the minimum amount of gas the transaction requires; commonly called **gas limit** or similar in our code and docs.

Author: The variable is nearly always called limit in the code, this is why I'm mentioning it, because it's darn confusing.

> 2. Call the [`zks_estimateGasL1ToL2`](../../api/api.md#zks_estimategasl1tol2) method, passing the transaction data. This returns the minimum amount of gas the transaction requires; commonly called **gas limit** or similar in our code and docs.

Author: i'm committing this change given the confusion

> 2. Call the [`zks_estimateGasL1ToL2`](../../api/api.md#zks_estimategasl1tol2) method, passing the transaction data. This returns the minimum amount of gas the transaction requires; commonly called **gas limit** or similar in our code and docs.

Reviewer 3: > The variable is nearly always called (something containing) "limit" in our code, inc. libraries and examples, this is why I'm mentioning it, because it's darn confusing.

Just to be clear, there is a variable that is called gasLimit (and it's the max).
The process of assigning a value to it (how do we know what should be the max?) is to call the estimateGas function (note, a function, not variable). This gives us the amount of gas we expect the tx to consume (in EVM we cannot predict it 100%, hence only estimation). Once we have this estimation, we multiply it be some factor (1.1, 1.3, 2, whatever) just to make sure that we don't send a tx with too little of gas, which will result in tx failure. And this would be the gasLimit we assign to this tx.

> 2. Call the [`zks_estimateGasL1ToL2`](../../api/api.md#zks_estimategasl1tol2) method, passing the transaction data. This returns the estimated amount of gas the transaction requires; commonly called **gas limit** or similar in our code and docs.

Reviewer: Just to make a clear distinction

```suggestion
2. Call the [`zks_estimateGasL1ToL2`](../../api/api.md#zks_estimategasl1tol2) method, passing the transaction data. This returns the estimated amount of L2 gas the transaction requires; commonly called **gas limit** or similar in our code and docs.
```

>     - Apply an alias to the addresses in the request if the sender address is a contract. If the sender is an EOA, no aliasing is required. This is implemented by the [`applyL1ToL2Alias`](https://github.com/matter-labs/zksync-2-contracts/blob/7b5c094a57c0606785ea38c9c752f9def9a5ed9d/ethereum/contracts/vendor/AddressAliasHelper.sol#L28) Solidity function.

Author: My feeling is link to somewhere as this is a how to so we're not explaining why things happen, just how to do things. I, personally, just accept the instruction. I don't think we have an explanation of this. Could you supply a paragraph @barakshani and I'll add it to the Interop page?

>     - Apply an alias to the addresses in the request if the sender address is a contract. If the sender is an EOA, no aliasing is required. This is implemented by the [`applyL1ToL2Alias`](https://github.com/matter-labs/zksync-2-contracts/blob/7b5c094a57c0606785ea38c9c752f9def9a5ed9d/ethereum/contracts/vendor/AddressAliasHelper.sol#L28) Solidity function.

Reviewer 3: No, as I don't hold this knowledge : )

>     - Apply an alias to the addresses in the request if the sender address is a contract. If the sender is an EOA, no aliasing is required. This is implemented by the [`applyL1ToL2Alias`](https://github.com/matter-labs/zksync-2-contracts/blob/7b5c094a57c0606785ea38c9c752f9def9a5ed9d/ethereum/contracts/vendor/AddressAliasHelper.sol#L28) Solidity function.

Author: OK, a few words on why aliasing @StanislavBreadless please?

>     - Apply an alias to the addresses in the request if the sender address is a contract. If the sender is an EOA, no aliasing is required. This is implemented by the [`applyL1ToL2Alias`](https://github.com/matter-labs/zksync-2-contracts/blob/7b5c094a57c0606785ea38c9c752f9def9a5ed9d/ethereum/contracts/vendor/AddressAliasHelper.sol#L28) Solidity function.

Author: I think this is a separate docs ticket.

>     - Apply an alias to the addresses in the request if the sender address is a contract. If the sender is an EOA, no aliasing is required. This is implemented by the [`applyL1ToL2Alias`](https://github.com/matter-labs/zksync-2-contracts/blob/7b5c094a57c0606785ea38c9c752f9def9a5ed9d/ethereum/contracts/vendor/AddressAliasHelper.sol#L28) Solidity function.

Reviewer 3: sure, whatever you find fit.
but please:
1. add a todo in the text (we already have 1 todo in this PR)
2. open the ticket
3. on the ticket, mention the todo

>     - Apply an alias to the addresses in the request if the sender address is a contract. If the sender is an EOA, no aliasing is required. This is implemented by the [`applyL1ToL2Alias`](https://github.com/matter-labs/zksync-2-contracts/blob/7b5c094a57c0606785ea38c9c752f9def9a5ed9d/ethereum/contracts/vendor/AddressAliasHelper.sol#L28) Solidity function.

Author: https://linear.app/matterlabs/issue/DOC-356/explain-aliasing -> no todo as it won't be in this doc

>     - Apply an alias to the addresses in the request if the sender address is a contract. If the sender is an EOA, no aliasing is required. This is implemented by the [`applyL1ToL2Alias`](https://github.com/matter-labs/zksync-2-contracts/blob/7b5c094a57c0606785ea38c9c752f9def9a5ed9d/ethereum/contracts/vendor/AddressAliasHelper.sol#L28) Solidity function.

Reviewer 3: TODO: add a link to aliasing

3. Call the [`l2TransactionBaseCost`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/interfaces/IMailbox.sol#L129) function, passing the gas price and gas limit from previous steps. This function returns the base cost required to send a transaction.

    - You also need to pass the constant representing how much gas is required to publish a byte of data from L1 to L2. At the time of writing, the JavaScript API provides this constant as [`REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_LIMIT`](../../api/js/utils.md#gas).

Once you have all of the above, you can send a transaction with the [`requestL2Transaction`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/interfaces/IMailbox.sol#L119) function, passing the gas limit returned at step 2, along with the gas price and base cost from steps 1 and 3 as part of the value parameters.

For a more detailed explanation on estimating gas, with code examples, find out [how to send a transaction from L1 to L2](../how-to/send-transaction-l1-l2.md).

## L2 to L2

> Gas estimation for L2 to L2 transactions on zkSyncEra works in the same way as it does in Ethereum.

Reviewer: ```suggestion
Gas estimation for L2 to L2 transactions on zkSync Era works in the same way as it does in Ethereum.
```

> Supply the same call data as an L1 to L2 transaction, use the Ethereum `eth_estimateGas` method, and send the transaction to an L2 node.

Reviewer 3: I don't understand your comment
> Supply the same call data as an L1 to L2 transaction, use the Ethereum `eth_estimateGas` method, and send the transaction to an L2 node.

This is "what", not "why".

> Also, which bits in the previous section were unnecessary?

They are unnecessary for the reader who is interested in L2-L2 transactions.

> Supply the same call data as an L1 to L2 transaction, use the Ethereum `eth_estimateGas` method, and send the transaction to an L2 node.

Author: Like code, we shouldn't duplicate anything.

> Supply the same call data as an L1 to L2 transaction, use the Ethereum `eth_estimateGas` method, and send the transaction to an L2 node.

Author: this is how...which is what we want "Supply the same call data as an L1 to L2 transaction, use the Ethereum eth_estimateGas method, and send the transaction to an L2 node."

Lets have a chat about this tomorrow morning if you're around @barakshani cos I'm not getting you :D

> Supply the same call data as an L1 to L2 transaction, use the Ethereum `eth_estimateGas` method, and send the transaction to an L2 node.

Reviewer 3: Yes, it's how, I agree.
All I'm saying is that I believe that most readers would be interested in the L2-L2 part. So directing them to the L1-L2 isn't very helpful.

In this case I would duplicate the similar things, but if not then my suggestion is to fully explain the L2-L2 part, then direct to it at the L1-L2 part.

> Supply the same call data as an L1 to L2 transaction, use the Ethereum `eth_estimateGas` method, and send the transaction to an L2 node.

Author: This sounds like personal preference. I'm not sure what is missing from this section tbh.

> Supply the same call data as an L1 to L2 transaction, use the Ethereum `eth_estimateGas` method, and send the transaction to an L2 node.

Reviewer 3: your call

> Supply the same call data as an L1 to L2 transaction, use the Ethereum `eth_estimateGas` method, and send the transaction to an L2 node.

Reviewer: Just a convention

```suggestion
Supply the same calldata as an L1 to L2 transaction, use the Ethereum `eth_estimateGas` method, and send the transaction to an L2 node.
```

:::info More info
For more information and live testing, check out the [Ethereum JSON RPC docs](https://ethereum.github.io/execution-apis/api-documentation/).
:::

## L2 to L1

> Gas estimation is dealt with implicitly by the SDK for L2 to L1 transactions.

Reviewer: Better to say that all the L2->L1 messages are parts of the corresponding L2->L2 transactions

Author: Remove mention of gas estimation?
Or something like **gas estimation for L2->L1 messages is dealt with by the corresponding L2->L2 txs** ??

Author: @barakshani ?

Reviewer 3: I'm not the one who made the initial comment.

Author: All input is welcome to get this out the door.

Author: @StanislavBreadless are you happy with the bolded sentence above?

Reviewer: @kailash-manasarovar yes
